sycl/source/kernel.cpp

```diff
@@ -6,13 +6,83 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <CL/sycl/detail/kernel_impl.hpp>
 #include <CL/sycl/kernel.hpp>
-
 #include <CL/sycl/program.hpp>
 
 namespace cl {
 namespace sycl {
 
+kernel::kernel(cl_kernel ClKernel, const context &SyclContext)
+    : impl(std::make_shared<detail::kernel_impl>(
+          detail::pi::cast<detail::RT::PiKernel>(ClKernel), SyclContext)) {}
+
+bool kernel::operator==(const kernel &RHS) const { return impl == RHS.impl; }
+
+bool kernel::operator!=(const kernel &RHS) const { return !operator==(RHS); }
```

Comment:
Suggest moving to the header.

Reply:
@romanovvlad impl is incomplete type in kernel.hpp, thus I can't move L20 to header.

Reply:
AFAIK you don't need kernel_impl to be complete type in kernel.hpp to just compare pointers.